File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        p1 = head
        p2 = head.next

        if p2 is None:
            return False

        while p2 and p2.next != None:
            if p1 == p2:
                return True
            p2 = p2.next.next
            p1 = p1.next
        
        return False

File: Data Structures & Algorithms/add-two-numbers/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = new ListNode(0);
        ListNode pr = res;
        int carry = 0;

        while (l1 != null || l2 != null) {
            int v1 = (l1 != null) ? l1.val : 0;
            int v2 = (l2 != null) ? l2.val : 0;

            int v3 = (v1 + v2 + carry) % 10;
            carry = (v1 + v2 + carry) / 10;
            pr.next = new ListNode(v3);
            pr = pr.next;

            l1 = (l1 != null) ? l1.next : null;
            l2 = (l2 != null) ? l2.next : null;
        }

        if (carry != 0) {
            pr.next = new ListNode(carry);
        }

        return res.next;
    }
}
